### PR TITLE
[SPARK-50584][SS] Modify state eviction metrics in TransformWithState to represent physical deletions

### PR DESCRIPTION
**Pull request by neilramaswamy**
### What changes were proposed in this pull request?

These changes ensure that TWS state variables treat `numRowsRemoved` as a physical metric representing the number of entries that they cleared out of the state store in response to TTL expiry, or a user action like `clear()`.

Note: `numRowsRemoved`, with these changes, does not track every single _deletion_ to RocksDB. For example, if we delete a row and then re-insert it, we don't count that as a removal. A removal is just about the number of records removed due to TTL expiry or a user clearing it.

Also, we were only using the count index to populate the `numRowsRemoved` metric; now that this metric is all physical, we don't need the count index.

### Why are the changes needed?

The current metrics are confusing. `numRowsRemoved` is hard to reconcile with `numTotalRows`; today it's even possible for the sum of `numRowsRemoved` across batches to be greater than `numTotalRows`.

### Does this PR introduce _any_ user-facing change?

Yes, but just with metrics.

### How was this patch tested?

Modified existing metrics in UTs to have the new behavior, and added more logic (and assertions) to all the suites to test the new behavior.

### Was this patch authored or co-authored using generative AI tooling?

Github Copilot

### CODE REVIEW

**Comment by neilramaswamy** [comment]
cc: @anishshri-db, I'm realizing that if `numRowsRemoved` is about physical rows, then when a user `clear`s their list state, we no longer need the entry count index to tell us how many things were removed. It will just show up as one physical row removed. If you're fine with that, I'll remove all the entry count logic.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala [hunk 1]
```diff
@@ -81,8 +81,8 @@ class ProgressReporter(
 
     addNewProgress(newProgress)
     postEvent(new QueryProgressEvent(newProgress))
-    logInfo(
-      log"Streaming query made progress: ${MDC(LogKeys.STREAMING_QUERY_PROGRESS, newProgress)}")
+    println(
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala:84]: Prob need to revert this ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala [hunk 2]
```diff
@@ -279,15 +279,14 @@ abstract class OneToOneTTLState(
     var numValuesExpired = 0L
 
     ttlEvictionIterator().foreach { ttlKey =>
-      // Delete from secondary index
-      deleteFromTTLIndex(ttlKey)
-      // Delete from primary index
       store.remove(toTTLRow(ttlKey).elementKey, stateName)
+      deleteFromTTLIndex(ttlKey)
+
+      TWSMetricsUtils.incrementMetric(metrics, "numRemovedStateRows", 2L)
```
[the remark this answers is not in the document]
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:285]: Hm, do you not think it's clear from the `store.remove` call above and `deleteFromTTLIndex`?

[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:285]: I mean - better to make it explicit - to help someone new understand the code ?

**Review by anishshri-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (neilramaswamy) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala:285]: Ok, let's see what Jungtaek says—I think he has mentioned before to not be super verbose in comments. I can add it if he doesn't have an opinion.

**Comment by HeartSaVioR** [comment]
I'm actually confused about the definition of "physical" here. If we define "physical row" as actual key-value in RocksDB, every single timer has to be considered and also every secondary index has to be considered. For this case I'm not very sure how users could consume the information.

In stream-stream join, it only counts the number of input rows, ignoring the meta rows. It only counts from the `keyWithIndexToValue` state store, excluding `keyToNumValues` for the count. memory is accounted for both sides.

That said, I'd lean on defining the "logical" rows better. If we consider a list having 100s of elements as "a" row, a map having 100s of user keys should be also considered as "a" row. Also, we need to ask ourselves, if we provide the total number of rows, what is the intuitive understanding from moderate users? If we provide the value as the sum of (data row + TTL index + timers), would users be able to reason about the number? My honest voice is that unless we provide these values separately, they would mostly care about data rows.

I understand that knowing the actual number of key-values in RocksDB could be quite useful on investigation or so, but if the number is hard to reason about from moderate users, we are required to provide these numbers separately anyway. I'd use the memory portion (though we know there is an issue on capturing RocksDB memory usage), or name the metric clearer to represent that the number is about "physical" key values and leave the existing metrics as they are (logical rows).

**Comment by neilramaswamy** [comment]
> if we provide the total number of rows

We do this, right? Today, the total number of rows is the _physical_ number of entries in RocksDB. If you had 100s of list elements, you'd have 4 `numTotalRows` (1 primary, 3 secondary indexes), but then 100 `numRowsRemoved`.

I think that is confusing. We should make it consistent—either make `numRowsTotal` as the number of primary entries, or make `numRowsRemoved` the physical. I personally have a preference for the latter (as done in this PR), and I would even argue that we should modify s/s joins to reflect this.

Then, the custom metrics like `numValues...` that _only_ the streaming physical operator knows about can be populated by the streaming physical operator.

What do you think?

**Comment by neilramaswamy** [comment]
Chatted with @HeartSaVioR about this offline, and here's the realizations that we had:

- `numTotalRows` is broken for TWS because this counting happens in `RocksDB.scala`, which counts _all_ rows across _all_ virtual column families.
- So, this is counting the number of physical rows, even though other operators count the number of logical rows. For example, stream-stream joins adds `1` to `numTotalRows` when one key is present, even though under the hood two state stores are used to hold it.
- Ideally, all metrics are logical, since users shouldn't be having to read metrics about Structured Streaming's implementation details, which may change a lot between releases.
- Physical metrics can be surfaced in `customMetrics`.

I agree with all of these points, so for now, we'll leave `numRowsRemoved` as is, as something that counts the number of _logical_ entries removed from state variables. It is future work to fix `numRowsTotal`. As a result, I'll be closing this PR.

**Pull request closed without merging**